### PR TITLE
Remove cast of string literal to string

### PR DESCRIPTION
**Pull request by DoctorKrolic**
Fixes: https://github.com/dotnet/roslyn/issues/65517

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs [hunk 1]
```diff
@@ -461,7 +461,11 @@ private static bool CastRemovalCouldCauseSignExtensionWarning(ExpressionSyntax c
                 return true;
 
             // We can safely remove convertion to object in interpolated strings regardless of nullability
-            if (castNode.IsParentKind(SyntaxKind.Interpolation) && originalConversionOperation.Type?.SpecialType is SpecialType.System_Object)
+            if (castNode.IsParentKind(SyntaxKind.Interpolation) && originalConvertedType.SpecialType is SpecialType.System_Object)
+                return true;
+
+            // Can remove convertion of string literal to string regardless of nullability
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs:467]: Convertion

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs [hunk 2]
```diff
@@ -461,7 +461,11 @@ private static bool CastRemovalCouldCauseSignExtensionWarning(ExpressionSyntax c
                 return true;
 
             // We can safely remove convertion to object in interpolated strings regardless of nullability
-            if (castNode.IsParentKind(SyntaxKind.Interpolation) && originalConversionOperation.Type?.SpecialType is SpecialType.System_Object)
+            if (castNode.IsParentKind(SyntaxKind.Interpolation) && originalConvertedType.SpecialType is SpecialType.System_Object)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs:464]: Can you add an assertion that these two types are the same?

File: src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs [hunk 3]
```diff
@@ -12621,5 +12620,33 @@ void Goo()
 ";
             await VerifyCS.VerifyCodeFixAsync(code, code);
         }
+
+        [Theory, WorkItem(65517, "https://github.com/dotnet/roslyn/issues/65517")]
+        [InlineData("string")]
+        [InlineData("string?")]
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs:12626]: Can you add test where there is `string?` but it affects some other type being created (like a `List<string?>`) through generic type inference. We don't want that inference to now create a different type.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs [hunk 4]
```diff
@@ -461,7 +461,11 @@ private static bool CastRemovalCouldCauseSignExtensionWarning(ExpressionSyntax c
...
+                return true;
+
+            // Can remove convertion of string literal to string regardless of nullability
+            if (rewrittenExpression.WalkDownParentheses().IsKind(SyntaxKind.StringLiteralExpression) && originalConvertedType.SpecialType is SpecialType.System_String)
```
Comment (Youssef1313) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs:468]: Why is `string` special here? Unless I'm missing something, I think this needs a generalized approach, for example, `var c = new C(); M(c);`

Comment (DoctorKrolic) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs:468]: Now when you've said it I am wondering if it is ok to simplify any cast of non-nullable to nullable when the type is a reference type🤔

File: src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs (hunk 3 above)
Comment (DoctorKrolic) [reply src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs:12626]: Can you give a code sample where removing the cast can mess up things. I cannot come up with something, where removing the cast can create nullability issues

Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs:12626]:
```c#    List<string?> GetListOfThings()
    {
        return CreateList((string?)"");
    }

    List<T> CreateList<T>(T value) where T : class? => new List<T> { value };
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs (hunk 1 above)
Comment (DoctorKrolic) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs:467]: What is the issue here? I don't see anything misspelled

Comment (Youssef1313) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs:467]: Should be "conversion" I think.

File: src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs (hunk 3 above)
Comment (DoctorKrolic) [reply src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs]: Added logic to handle this specific case. However, if I extract the string into a separate variable, something weird happens: https://github.com/dotnet/roslyn/issues/65520. Not sure if it is correct behaviour, so no tests for it so far

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs (hunk 1 above)
Comment (DoctorKrolic) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs]: Uh, yes, thanks @Youssef1313!

**Pull request closed without merging**